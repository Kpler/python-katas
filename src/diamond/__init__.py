def get_diamond(letter: str):

    return letter
