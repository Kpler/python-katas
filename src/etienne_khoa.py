class Employee:
    _frozen = False

    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name
        self._frozen = True
#
    # def __setattr__(self, *args, **kwargs):
    #     # if attr in ['first_name', ]:  # 'last_name'
    #     if self._frozen:
    #         raise AttributeError('error')
    #     object.__setattr__(self, *args, **kwargs)

    def __str__(self):
        return f'my name is {self.last_name}, {self.first_name} {self.last_name}'


if __name__ == '__main__':
    toto = Employee('Elon', "D'haussy")
    print(toto.first_name)
    print(toto)
    toto.last_name = 'SF'
    # toto.first_name = 'SF'
    print(toto)
