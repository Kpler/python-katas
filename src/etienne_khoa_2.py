class Human:
    def __init__(self, first_name):
        self.first_name = first_name


class Employee(Human):
    def __init__(self, first_name, last_name):
        super().__init__(first_name)
        self.last_name = last_name

    def __str__(self):
        return f'my name is {self.last_name}, {self.first_name} {self.last_name}'


if __name__ == '__main__':
    toto = Employee('Elon', "D'haussy")
    print(toto)
