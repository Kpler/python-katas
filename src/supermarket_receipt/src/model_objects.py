from __future__ import annotations
from enum import Enum


class Product:
    def __init__(self, name, unit: ProductUnit):
        self.name = name
        self.unit = unit


class ProductQuantity:
    def __init__(self, product, quantity: float):
        self.product = product
        self.quantity = quantity


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


class SpecialOfferType(Enum):
    THREE_FOR_TWO = 1
    TEN_PERCENT_DISCOUNT = 2
    TWO_FOR_AMOUNT = 3
    FIVE_FOR_AMOUNT = 4


class Offer:
    def __init__(self, offer_type: SpecialOfferType, product: Product, argument: float) -> None:
        self.offer_type = offer_type
        self.product = product
        self.argument = argument


class Discount:
    def __init__(self, product: Product, description: str, discount_amount: float) -> None:
        self.product = product
        self.description = description
        self.discount_amount = discount_amount
