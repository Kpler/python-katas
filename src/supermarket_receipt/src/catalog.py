from src.model_objects import Product


class SupermarketCatalog:
    def __init__(self):
        self.products: dict[str, Product] = {}
        self.prices: dict[str, float] = {}

    def add_product(self, product: Product, price: float) -> None:
        raise Exception("cannot be called from a unit test - it accesses the database")

    def unit_price(self, product: Product) -> float:
        raise Exception("cannot be called from a unit test - it accesses the database")
