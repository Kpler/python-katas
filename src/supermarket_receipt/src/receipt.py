from src.model_objects import Discount, Product


class ReceiptItem:
    def __init__(self, product: Product, quantity: float, price: float, total_price: float) -> None:
        self.product = product
        self.quantity = quantity
        self.price = price
        self.total_price = total_price


class Receipt:
    def __init__(self) -> None:
        self._items: list[ReceiptItem] = []
        self._discounts: list[Discount] = []

    def total_price(self) -> float:
        total = 0
        for item in self.items:
            total += item.total_price
        for discount in self.discounts:
            total += discount.discount_amount
        return total

    def add_product(self, product: Product, quantity: float, price: float, total_price: float) -> None:
        self._items.append(ReceiptItem(product, quantity, price, total_price))

    def add_discount(self, discount: Discount) -> None:
        self._discounts.append(discount)

    @property
    def items(self) -> list[ReceiptItem]:
        return self._items[:]

    @property
    def discounts(self) -> list[Discount]:
        return self._discounts[:]
