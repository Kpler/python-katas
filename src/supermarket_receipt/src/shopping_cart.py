import math

from supermarket_receipt.src.model_objects import ProductQuantity, SpecialOfferType, Discount


class ShoppingCart:

    def __init__(self):
        self._items = []
        self._product_quantities = {}

    @property
    def items(self):
        return self._items

    def add_item(self, product):
        self.add_item_quantity(product, 1.0)

    @property
    def product_quantities(self):
        return self._product_quantities

    def add_item_quantity(self, product, quantity):
        self._items.append(ProductQuantity(product, quantity))
        if product in self._product_quantities.keys():
            self._product_quantities[product] = self._product_quantities[product] + quantity
        else:
            self._product_quantities[product] = quantity

    def handle_offers(self, receipt, offers, catalog):
        for p in self._product_quantities.keys():
            quantity = self._product_quantities[p]
            if p in offers.keys():
                offer = offers[p]
                unit_price = catalog.unit_price(p)
                quantity_as_int = int(quantity)
                discount = None
                x = 1
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO:
                    x = 3

                elif offer.offer_type == SpecialOfferType.TWO_FOR_AMOUNT:
                    x = 2
                    if quantity_as_int >= 2:
                        total = offer.argument * (quantity_as_int / x) + quantity_as_int % 2 * unit_price
                        discount_n = unit_price * quantity - total
                        discount = Discount(p, "2 for " + str(offer.argument), -discount_n)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT:
                    x = 5

                number_of_x = math.floor(quantity_as_int / x)
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO and quantity_as_int > 2:
                    discount_amount = quantity * unit_price - (
                                (number_of_x * 2 * unit_price) + quantity_as_int % 3 * unit_price)
                    discount = Discount(p, "3 for 2", -discount_amount)

                if offer.offer_type == SpecialOfferType.TEN_PERCENT_DISCOUNT:
                    discount = Discount(p, str(offer.argument) + "% off",
                                        -quantity * unit_price * offer.argument / 100.0)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT and quantity_as_int >= 5:
                    discount_total = unit_price * quantity - (
                                offer.argument * number_of_x + quantity_as_int % 5 * unit_price)
                    discount = Discount(p, str(x) + " for " + str(offer.argument), -discount_total)

                if discount:
                    receipt.add_discount(discount)

    def handle_bundles(self, receipt, bundle_offers, catalog):
        """
        if products of bundle_offers are in the cart, apply the discount
        """
        for bundle in bundle_offers.keys():
            discount = 0

            for product in bundle:
                if product in self._product_quantities.keys():
                    quantity = self._product_quantities[product]
                    unit_price = catalog.unit_price(product)
                    discount += quantity*unit_price * bundle_offers[bundle].discount_percentage / 100.0

            if discount > 0:
                bundle_offer = bundle_offers[bundle]
                discount = Discount(product, str(bundle_offer.discount_percentage) + " discount on bundle", -discount)
                receipt.add_discount(discount)