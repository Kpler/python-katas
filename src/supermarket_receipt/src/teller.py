import math

from src.catalog import SupermarketCatalog
from src.model_objects import Offer, Product, SpecialOfferType, Discount
from src.receipt import Receipt
from src.shopping_cart import ShoppingCart

OfferMap = dict[Product, Offer]


class Teller:
    """The entity scanning the item."""

    def __init__(self, catalog: SupermarketCatalog) -> None:
        self.catalog: SupermarketCatalog = catalog
        self.offers: OfferMap = {}

    def add_special_offer(self, offer_type: SpecialOfferType, product: Product, argument: float):
        self.offers[product] = Offer(offer_type, product, argument)

    @staticmethod
    def handle_offers(cart, offers, catalog):
        discounts: list[Discount] = []
        for p in cart._product_quantities.keys():
            quantity = cart._product_quantities[p]
            if p in offers.keys():
                offer = offers[p]
                unit_price = catalog.unit_price(p)
                quantity_as_int = int(quantity)
                discount = None
                x = 1
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO:
                    x = 3

                elif offer.offer_type == SpecialOfferType.TWO_FOR_AMOUNT:
                    x = 2
                    if quantity_as_int >= 2:
                        total = (
                            offer.argument * (quantity_as_int / x)
                            + quantity_as_int % 2 * unit_price
                        )
                        discount_n = unit_price * quantity - total
                        discount = Discount(p, "2 for " + str(offer.argument), -discount_n)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT:
                    x = 5

                number_of_x = math.floor(quantity_as_int / x)
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO and quantity_as_int > 2:
                    discount_amount = quantity * unit_price - (
                        (number_of_x * 2 * unit_price) + quantity_as_int % 3 * unit_price
                    )
                    discount = Discount(p, "3 for 2", -discount_amount)

                if offer.offer_type == SpecialOfferType.TEN_PERCENT_DISCOUNT:
                    discount = Discount(
                        p,
                        str(offer.argument) + "% off",
                        -quantity * unit_price * offer.argument / 100.0,
                    )

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT and quantity_as_int >= 5:
                    discount_total = unit_price * quantity - (
                        offer.argument * number_of_x + quantity_as_int % 5 * unit_price
                    )
                    discount = Discount(p, str(x) + " for " + str(offer.argument), -discount_total)

                if discount:
                    discounts.append(discount)

        return discounts

    def checks_out_articles_from(self, the_cart: ShoppingCart) -> Receipt:
        receipt = Receipt()

        product_quantities = the_cart.items
        for pq in product_quantities:
            p = pq.product
            quantity = pq.quantity
            unit_price = self.catalog.unit_price(p)
            price = quantity * unit_price
            receipt.add_product(p, quantity, unit_price, price)

        discounts = self.handle_offers(the_cart, self.offers, self.catalog)
        for discount in discounts:
            receipt.add_discount(discount)

        return receipt
