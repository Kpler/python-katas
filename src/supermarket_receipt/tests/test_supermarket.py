import pytest

from supermarket_receipt.src.model_objects import Product, SpecialOfferType, ProductUnit
from supermarket_receipt.src.shopping_cart import ShoppingCart
from supermarket_receipt.src.teller import Teller
from supermarket_receipt.tests.fake_catalog import FakeCatalog


def test_ten_percent_discount():
    catalog = FakeCatalog()
    teller = Teller(catalog)
    cart = ShoppingCart()

    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    catalog.add_product(toothbrush, 0.99)
    catalog.add_product(apples, 1.99)
    teller.add_special_offer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush, 10.0)

    cart.add_item_quantity(apples, 2.5)
    receipt = teller.checks_out_articles_from(cart)

    assert 4.975 == pytest.approx(receipt.total_price(), 0.01)
    assert [] == receipt.discounts
    assert 1 == len(receipt.items)
    receipt_item = receipt.items[0]
    assert apples == receipt_item.product
    assert 1.99 == receipt_item.price
    assert 2.5 * 1.99 == pytest.approx(receipt_item.total_price, 0.01)
    assert 2.5 == receipt_item.quantity

def test_explore():
    catalog = FakeCatalog()
    teller = Teller(catalog)
    cart = ShoppingCart()

    # things to catalog
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    toothpaste = Product("toothpaste", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    catalog.add_product(toothbrush, 1)
    catalog.add_product(toothpaste, 3)

    teller.add_bundle_discount([toothbrush, toothpaste], 10)

    # things to cart
    cart.add_item_quantity(toothbrush, 1)
    cart.add_item_quantity(toothpaste, 1)
    receipt = teller.checks_out_articles_from(cart)
    assert 3.6 == pytest.approx(receipt.total_price(), 0.01)



def test_explore_2():
    catalog = FakeCatalog()
    teller = Teller(catalog)
    cart = ShoppingCart()

    # things to catalog
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    toothpaste = Product("toothpaste", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    catalog.add_product(toothbrush, 1)
    catalog.add_product(toothpaste, 3)

    teller.add_bundle_discount([toothbrush, toothpaste], 10)

    # things to cart
    cart.add_item_quantity(toothbrush, 2)
    cart.add_item_quantity(toothpaste, 2)
    receipt = teller.checks_out_articles_from(cart)
    assert 7.2 == pytest.approx(receipt.total_price(), 0.01)


def test_explore_3():
    catalog = FakeCatalog()
    teller = Teller(catalog)
    cart = ShoppingCart()

    # things to catalog
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    toothpaste = Product("toothpaste", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    catalog.add_product(toothbrush, 1)
    catalog.add_product(toothpaste, 3)

    teller.add_bundle_discount([toothbrush, toothpaste], 10)

    # things to cart
    cart.add_item_quantity(toothbrush, 1) # total: 1 -> - 10 -> 0.9
    cart.add_item_quantity(toothpaste, 2) # total: 6 -> - 10 on 1 -> 3 + 2.7 = 5.7
    receipt = teller.checks_out_articles_from(cart)
    assert 6.6 == pytest.approx(receipt.total_price(), 0.01)

