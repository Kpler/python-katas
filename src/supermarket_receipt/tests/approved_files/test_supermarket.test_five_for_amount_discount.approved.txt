toothbrush                          4.95
  0.99 * 5
5 for 4.0 (toothbrush)             -0.95

Total:                              4.00
